README
Bing::OpenSearch version 0.0.2 - 07 February 2010

Bing::OpenSearch provides access to the OpenSearch compliant RSS interface of the Bing API version 2.0.
The RSS endpoint is anonymous; it doesn’t require an AppID.

However, the results returned may not be used, reproduced or transmitted in any manner or for any purpose other
than rendering Bing results within an RSS aggregator for your personal, non-commercial use. Any other use
requires written permission from Microsoft Corporation. By using these results in any manner whatsoever, you
agree to be bound by the foregoing restrictions.

INSTALLATION

perl Makefile.PL
make
make test
make install

DEPENDENCIES

- LWP::UserAgent
- XML::LibXML
- Class::Std::Utils
- version
- Encode
- Test::More
- Test::Exception
- Scalar::Util

COPYRIGHT AND LICENCE

Copyright (C) 2010, Kostas Ntonas <kntonas at gmail.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Bing::OpenSearch

0.0.1  Feb 05, 2010
       - Initial release.

0.0.2  Feb 07, 2010
       - Added 'save' and 'space' methods to Bing::OpenSearch::Response.
       - Minor changes to documentation and README.
       - Modified Terms of Use: ONLY within an RSS aggregator for personal, non-commercial use.


test.pl
#!/usr/local/bin/perl -w
use strict;
## By virtue of being named "test.pl", this program is automatically run
## via "make test".

use Bing::OpenSearch;

my $response = Bing::OpenSearch->Query( 'Larry Wall', {count => 1} );

my @Results = $response->results;

if (@Results == 1 and $Results[0]->link =~ m{^https?://}) {
    print "Bing::OpenSearch test passes!\n";
}
else {
    die "Bing::OpenSearch test failed: $@\n";
}


lib/Bing/OpenSearch.pm
package Bing::OpenSearch;

use strict;
use warnings;

use Carp qw( croak );
use Scalar::Util qw( looks_like_number );

use LWP::UserAgent;
use XML::LibXML;

use Bing::OpenSearch::Response;

use version; our $VERSION = qv('0.0.2');

our $agent;

sub Query {
    my ($self, $searchTerms, $params_ref) = @_;

    if ( ! $searchTerms ) {
        croak "A search query must be provided.";
    }

    $searchTerms =~s#\s*\++\s*#\+#g;
    $searchTerms =~s#\s+#%20#g;  # replace white space characters with %20

    my $sourceType = $$params_ref{sourceType} || 'Web';
    if ($sourceType ne 'Web' and $sourceType ne 'Image') {
        croak "The supported search spaces are Web and Image. Please enter a valid sourceType parameter.";
    }

    my $count = $$params_ref{count} || 10;
    if (! looks_like_number $count) {
        croak "The count parameter should be a positive integer.";
    }
    elsif ($count <= 0) {
        croak "The count parameter should be a positive integer.";
    }

    my $startIndex = $$params_ref{offset} || 0;
    if (! looks_like_number $startIndex) {
        croak "The offset parameter should be an integer (>=0).";
    }
    elsif ($startIndex < 0) {
        croak "The offset parameter should be >= 0.";
    }

    my $language = $$params_ref{market};

    my $url = "http://api.bing.com/rss.aspx?source=$sourceType&query=$searchTerms&$sourceType.count=$count&".
              "$sourceType.offset=$startIndex&FORM=MO0000&version=2.0";

    if ($language) {
        $url .= "&market=$language";
    }

    $agent ||= LWP::UserAgent->new(agent => "Bing::OpenSearch ($Bing::OpenSearch::VERSION)", env_proxy  => 1);

    my $response = $agent->get($url);
    unless ($response->is_success) { croak "Failed to fetch $url"; }

    return Bing::OpenSearch::Response->new($response->content, $url, $sourceType);
}

1;
__END__

=head1 NAME

Bing::OpenSearch - Access the OpenSearch compliant RSS interface of the Bing API version 2.0.

=head1 VERSION

Version 0.0.2

=head1 SYNOPSIS

 use strict;
 use warnings;

 use Bing::OpenSearch;

 # Query Bing Image search space for 'Greece'
 my $response = Bing::OpenSearch->Query( 'Greece', { sourceType => 'Image',
                                                     offset => 0,
                                                     count => 10,
                                                     market => 'en-us'
                                                   }
                                       );

 print "Query URL: ", $response->queryURL,"\n";
 print "Number of Results: ", $response->totalResults,"\n";

 for my $item ($response->results) { # parse results and print data
    print "title: ", $item->title,"\n";
    print "link : ", $item->link,"\n";
    print "guid : ", $item->guid,"\n";
    if ($item->description) {
        print "description: ", $item->description,"\n";
    }
    if ($item->pubDate) {
        print "pubDate: ", $item->pubDate,"\n";
    }

    if ($response->space eq 'Image') { # ONLY for Image search space:
        print "media-content-url     : ", $item->media_content->url,"\n";
        print "media-content-height  : ", $item->media_content->height,"\n";
        print "media-content-width   : ", $item->media_content->width,"\n";
        print "media-content-fileSize: ", $item->media_content->fileSize,"\n";
        print "media-content-type    : ", $item->media_content->type,"\n";
        print "media-thumbnail-url   : ", $item->media_thumbnail->url,"\n";
        print "media-thumbnail-height: ", $item->media_thumbnail->height,"\n";
        print "media-thumbnail-width : ", $item->media_thumbnail->width,"\n";
    }
    print "\n";
 }

 # save the response feed to an RSS UTF-8 file
 $response->save('bing_response.rss');

=head1 DESCRIPTION

Bing::OpenSearch facilitates searching Bing through its API Version 2.0 via the OpenSearch compliant RSS interface.
The RSS endpoint is anonymous; it doesn't require an AppID.

The following search spaces (sourceTypes) are supported:

=over

=item * I<Web> - searches for web content.

=item * I<Image> - searches for images on the web.

=back

=head1 INTERFACE

=head2 Query( $searchTerms [, \%params] )

Searches Bing for the given query ($searchTerms) using the given search parameters and
returns a L<Bing::OpenSearch::Response> object.

Valid search parameters include:

=over

=item * I<sourceType> - search space, default is Web.

=item * I<offset> - startIndex, default is 0. Indicates how far into the result set
you are currently processing.

=item * I<count> - number of results requested, default is 10. The maximum number of results returned
from Bing API version 2.0 is 50. However, if you want more results, you can submit multiple queries
and utilize the offset parameter.

=item * I<market> - language code such as en-us (English), el-gr (Greek), es-es (Spanish), etc.
If provided, attempts to restrict the results to those in the given language.

=back

Note that all of these parameters are optional.

=head1 DEPENDENCIES

=over

=item * XML::LibXML

=item * LWP::UserAgent

=item * Scalar::Util

=item * Class::Std::Utils

=item * Encode

=item * version

=item * Test::More

=item * Test::Exception

=back

=head1 Terms of Use

The results returned may not be used, reproduced or transmitted in any manner or for any purpose other
than rendering Bing results within an RSS aggregator for your personal, non-commercial use. Any other use
requires written permission from Microsoft Corporation. By using these results in any manner whatsoever, you
agree to be bound by the foregoing restrictions.

=head1 BUGS AND SUGGESTIONS

Any feedback is very welcome. Please contact me for any bugs, comments or suggestions.

=head1 ACKNOWLEDGEMENTS

Special thanks to Petr Pajas, Damian Conway, Gisle Aas, Andy Lester, Jeffrey Friedl, Michael G Schwern,
Dan Kogai, Adrian Howard, John Peacock, David A P Mitchell and Graham Barr for their great modules and
their huge contribution to the Perl community. And most of all to Larry Wall, the man who started it all.

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2010, Kostas Ntonas, C<<kntonas at gmail.com>>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Bing/OpenSearch/Response.pm
package Bing::OpenSearch::Response;

use strict;
use warnings;

use Carp qw( croak );

use XML::LibXML;
use Class::Std::Utils;
use Encode;

use Bing::OpenSearch::Result;

use version; our $VERSION = qv('0.0.2');

# We use inside-out objects.
# See Chapters 15 and 16 of "Perl Best Practices" (O'Reilly, 2005) for details.
# Objects of this class have the following attributes
my %totalResults_of;
my %queryURL_of;
my %content_of;
my %results_of;
my %space_of;

sub new
{
    my ($class, $xml, $url, $space) = @_;

    my $response={};
    bless $response, $class;

    # Convert an object reference into a unique ID number
    my $id_num = ident $response;

    $queryURL_of{$id_num} = $url;
    $content_of{$id_num} = $xml;

    my $parser = XML::LibXML->new();

    # turn the parser warnings on and turn the parser's recover mode off
    $parser->recover_silently(0);

    # do not maintain white-space in the document
    $parser->keep_blanks(0);

    my $doc; # XML::LibXML::Document object, which is a DOM object

    eval { $doc = $parser->parse_string($xml) };

    if ($@) {
        croak "XML::LibXML parser error: $@";
    }

    $doc->indexElements(); # This function causes libxml2 to stamp all
    # elements in a document with their document position index which
    # considerably speeds up XPath queries for large documents.

    $totalResults_of{$id_num} = 0;
    if ($doc->getElementsByTagName('os:totalResults')) {
        $totalResults_of{$id_num} = ${$doc->getElementsByTagName('os:totalResults')}[0]->textContent;
    }

    $space_of{$id_num} = $space;

    my $itemList = $doc->getElementsByTagName('item'); # returns a XML::LibXML::NodeList object (SCALAR context)
    my @items;
    if ($itemList) {
        @items = $itemList->get_nodelist;
    }

    my @results;
    for my $item (@items) {
        my $result = Bing::OpenSearch::Result->new($item);
        push @results, $result;
    }

    $results_of{$id_num} = \@results;

    return $response;
}

sub DESTROY {
    my $self = shift;

    my $id_num = ident $self;

    delete $totalResults_of{$id_num};
    delete $queryURL_of{$id_num};
    delete $content_of{$id_num};
    delete $results_of{$id_num};
    delete $space_of{$id_num};
}

# get content
sub content {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $content_of{ident $self};
}

# get queryURL
sub queryURL {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $queryURL_of{ident $self};
}

# get totalResults
sub totalResults {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $totalResults_of{ident $self};
}

# get results
sub results {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    return @{$results_of{ident $self}} if wantarray; # list context - return array of results
    return $results_of{ident $self}; # scalar context - return reference to the array of results
}

# get search space from which the results derived
sub space {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $space_of{ident $self};
}

# save the response RSS feed to a UTF-8 file
sub save {
    my ($self, $filename) = @_;
    ref $self or croak "Instance variable needed.\n";

    open my $fh, '>:utf8', $filename;
    print $fh decode_utf8($self->content);
    close $fh;
}

1;
__END__

=head1 NAME

Bing::OpenSearch::Response - Container object for the result set of one query to the Bing OpenSearch interface.

=head1 VERSION

Version 0.0.2

=head1 PACKAGE USE

You don't need to C<use> this package directly.

=head1 CONSTRUCTOR

=head2 new( $xml, $url, $space )

Constructs a new instance of Bing::OpenSearch::Response using the RSS response feed
content ($xml), the query URL ($url) built for this particular query and the search
space queried ($space). You don't need to use this method.

=head1 METHODS

=head2 content

Returns the raw content of the RSS response feed.

=head2 queryURL

Returns the query URL built for the given search query.

=head2 results

Returns an array of L<Bing::OpenSearch::Result> objects if called in a list context.
If called in a scalar context, it returns a reference to the array of results.

=head2 totalResults

Returns the number of hits found.

=head2 space

Returns the search space from which the results derived.

=head2 save

Save the response RSS feed to a UTF-8 file.

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2010, Kostas Ntonas, C<<kntonas at gmail.com>>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Bing/OpenSearch/MediaContent.pm
package Bing::OpenSearch::MediaContent;

use strict;
use warnings;

use Carp qw( croak );

use XML::LibXML;
use Class::Std::Utils;

use version; our $VERSION = qv('0.0.2');

# We use inside-out objects.
# See Chapters 15 and 16 of "Perl Best Practices" (O'Reilly, 2005) for details.
# Objects of this class have the following attributes
my %url_of;
my %height_of;
my %width_of;
my %fileSize_of;
my %type_of;

sub new
{
    my $class = shift;
    my $node = shift; # XML::LibXML::Element object - media:content sub-element

    my $object={};
    bless $object, $class;

    # Convert an object reference into a unique ID number
    my $id_num = ident $object;

    $url_of{$id_num} = $node->getAttribute('url');
    $height_of{$id_num} = $node->getAttribute('height');
    $width_of{$id_num} = $node->getAttribute('width');
    $type_of{$id_num} = $node->getAttribute('type');
    $fileSize_of{$id_num} = $node->getAttribute('fileSize');

    return $object;
}

sub DESTROY {
    my $self = shift;

    my $id_num = ident $self;

    delete $url_of{$id_num};
    delete $height_of{$id_num};
    delete $width_of{$id_num};
    delete $fileSize_of{$id_num};
    delete $type_of{$id_num};
}

# get url
sub url {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $url_of{ident $self};
}

# get height
sub height {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $height_of{ident $self};
}

# get width
sub width {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $width_of{ident $self};
}

# get fileSize
sub fileSize {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $fileSize_of{ident $self};
}

# get type
sub type {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $type_of{ident $self};
}

1;
__END__

=head1 NAME

Bing::OpenSearch::MediaContent - Class representing a media:content sub-element of an item element of the Bing OpenSearch compliant response RSS feed.

=head1 VERSION

Version 0.0.2

=head1 PACKAGE USE

You don't need to C<use> this package directly.

=head1 DESCRIPTION

A MediaContent object represents a media:content sub-element of the response RSS feed and provides access methods for
all its attributes.

=head1 CONSTRUCTOR

=head2 new( $node )

Constructs a new instance of Bing::OpenSearch::MediaContent using $node,
which is a XML::LibXML::Element object which represents a media:content
sub-element of the RSS response feed. You don't need to use this method.

=head1 METHODS

=head2 url

Returns the url attribute of a media:content sub-element.

=head2 height

Returns the height attribute of a media:content sub-element.

=head2 width

Returns the width attribute of a media:content sub-element.

=head2 fileSize

Returns the fileSize attribute of a media:content sub-element.

=head2 type

Returns the type attribute of a media:content sub-element.

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2010, Kostas Ntonas, C<<kntonas at gmail.com>>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Bing/OpenSearch/MediaThumbnail.pm
package Bing::OpenSearch::MediaThumbnail;

use strict;
use warnings;

use Carp qw( croak );

use XML::LibXML;
use Class::Std::Utils;

use version; our $VERSION = qv('0.0.2');

# We use inside-out objects.
# See Chapters 15 and 16 of "Perl Best Practices" (O'Reilly, 2005) for details.
# Objects of this class have the following attributes
my %url_of;
my %height_of;
my %width_of;

sub new
{
    my $class = shift;
    my $node = shift; # XML::LibXML::Element object - media:thumbnail sub-element

    my $object={};
    bless $object, $class;

    # Convert an object reference into a unique ID number
    my $id_num = ident $object;

    $url_of{$id_num} = $node->getAttribute('url');
    $height_of{$id_num} = $node->getAttribute('height');
    $width_of{$id_num} = $node->getAttribute('width');

    return $object;
}

sub DESTROY {
    my $self = shift;

    my $id_num = ident $self;

    delete $url_of{$id_num};
    delete $height_of{$id_num};
    delete $width_of{$id_num};
}

# get url
sub url {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $url_of{ident $self};
}

# get height
sub height {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $height_of{ident $self};
}

# get width
sub width {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $width_of{ident $self};
}

1;
__END__

=head1 NAME

Bing::OpenSearch::MediaThumbnail - Class representing a media:thumbnail sub-element of an item element of the Bing OpenSearch compliant response RSS feed.

=head1 VERSION

Version 0.0.2

=head1 PACKAGE USE

You don't need to C<use> this package directly.

=head1 CONSTRUCTOR

=head2 new( $node )

Constructs a new instance of Bing::OpenSearch::MediaThumbnail using $node,
which is a XML::LibXML::Element object which represents a media:thumbnail
sub-element of the RSS response feed. You don't need to use this method.

=head1 METHODS

=head2 url

Returns the url attribute of a media:thumbnail sub-element.

=head2 height

Returns the height attribute of a media:thumbnail sub-element.

=head2 width

Returns the width attribute of a media:thumbnail sub-element.

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2010, Kostas Ntonas, C<<kntonas at gmail.com>>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.


lib/Bing/OpenSearch/Result.pm
package Bing::OpenSearch::Result;

use strict;
use warnings;

use Carp qw( croak );

use XML::LibXML;
use Class::Std::Utils;

use Bing::OpenSearch::MediaContent;
use Bing::OpenSearch::MediaThumbnail;

use version; our $VERSION = qv('0.0.2');

# We use inside-out objects.
# See Chapters 15 and 16 of "Perl Best Practices" (O'Reilly, 2005) for details.
# Objects of this class have the following attributes
my %title_of;
my %link_of;
my %guid_of;
my %description_of;
my %pubDate_of;
my %media_content_of;
my %media_thumbnail_of;

sub new
{
    my $class = shift;
    my $node = shift; # XML::LibXML::Element object

    my $result={};
    bless $result, $class;

    # Convert an object reference into a unique ID number
    my $id_num = ident $result;

    if ($node->getElementsByTagName('title')) {
        $title_of{$id_num} = ${$node->getElementsByTagName('title')}[0]->textContent;
    }
    if ($node->getElementsByTagName('link')) {
        $link_of{$id_num} = ${$node->getElementsByTagName('link')}[0]->textContent;
    }
    if ($node->getElementsByTagName('guid')) {
        $guid_of{$id_num} = ${$node->getElementsByTagName('guid')}[0]->textContent;
    }
    if ($node->getElementsByTagName('description')) {
        $description_of{$id_num} = ${$node->getElementsByTagName('description')}[0]->textContent;
    }
    if ($node->getElementsByTagName('pubDate')) {
        $pubDate_of{$id_num} = ${$node->getElementsByTagName('pubDate')}[0]->textContent;
    }

    if ($node->getElementsByTagName('media:content')) {
        my $media_content = ${$node->getElementsByTagName('media:content')}[0];
        $media_content_of{$id_num} = Bing::OpenSearch::MediaContent->new($media_content);
    }
    if ($node->getElementsByTagName('media:thumbnail')) {
        my $media_thumbnail = ${$node->getElementsByTagName('media:thumbnail')}[0];
        $media_thumbnail_of{$id_num} = Bing::OpenSearch::MediaThumbnail->new($media_thumbnail);
    }

    return $result;
}

sub DESTROY {
    my $self = shift;

    my $id_num = ident $self;

    delete $title_of{$id_num};
    delete $link_of{$id_num};
    delete $guid_of{$id_num};
    delete $description_of{$id_num};
    delete $pubDate_of{$id_num};
    delete $media_content_of{$id_num};
    delete $media_thumbnail_of{$id_num};
}

# get title
sub title {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $title_of{ident $self};
}

# get link
sub link {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $link_of{ident $self};
}

# get guid
sub guid {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $guid_of{ident $self};
}

# get description
sub description {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $description_of{ident $self};
}

# get pubDate
sub pubDate {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $pubDate_of{ident $self};
}

# get media:content hash
sub media_content {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $media_content_of{ident $self};
}

# get media:thumbnail hash
sub media_thumbnail {
    my $self = shift;
    ref $self or croak "Instance variable needed.\n";
    $media_thumbnail_of{ident $self};
}

1;
__END__

=head1 NAME

Bing::OpenSearch::Result - Class representing a single result from a Bing OpenSearch query.

=head1 VERSION

Version 0.0.2

=head1 PACKAGE USE

You don't need to C<use> this package directly.

=head1 DESCRIPTION

A Result object is a single result from a Bing OpenSearch query that provides access methods for
all sub-elements of item elements of the RSS response feed.

=head1 CONSTRUCTOR

=head2 new( $item )

Constructs a new instance of Bing::OpenSearch::Response using a XML::LibXML::Element object ($item)
which represents an item element of the RSS response feed. You don't need to use this method.

=head1 METHODS

=head2 title

Returns the title of a result.

=head2 link

Returns the link of a result.

=head2 guid

Returns the guid of a result.

=head2 description

Returns the description of a result.

=head2 pubDate

Returns the pubDate of a result.

=head2 media_content

Returns the media:content of a result as a L<Bing::OpenSearch::MediaContent> object
(only for a result from an image search query).

=head2 media_thumbnail

Returns the media:thumbnail of a result as a L<Bing::OpenSearch::MediaThumbnail> object
(only for a result from an image search query).

=head1 LICENCE AND COPYRIGHT

Copyright (c) 2010, Kostas Ntonas, C<<kntonas at gmail.com>>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE
LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL,
OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE
THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
SUCH DAMAGES.
